sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -4601,6 +4601,231 @@ case class ArrayExcept(left: Expression, right: Expression) extends ArrayBinaryL
     newLeft: Expression, newRight: Expression): ArrayExcept = copy(left = newLeft, right = newRight)
 }
 
+@ExpressionDescription(
+  usage = "_FUNC_(x, pos, val) - Places val into index pos of array x (array indices start at 1)",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(1, 2, 3, 4), 5, 5);
+       [1,2,3,4,5]
```

Comment:
The behavior is different from snowflake.
The result should be [1,2,3,4,undefined,5]
I don't know if we should use undefined here too. cc @zhengruifeng

Reply:
yes, the indices start at 0, see the examples in https://github.com/apache/spark/pull/38867#issuecomment-1365013704

Reply:
All array functions of Spark follow a 1-base index to my understanding; for example, many functions of the collectionOperations.scala have something like the following comment:
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala#L2080

To me, it sounds like a bad idea to be inconsistent between these new functions and legacy functions. I'm happy to amend the code if we want to follow a 0 based index however (but I do think people would likely make mistakes in their code if we implemented array_insert as the only array function using a 0 index)

Reply:
Yes. This is an issue indeed before we discuss how to do it.
These legacy functions let users know each array function starts at 0, if we change the new functions based on 1, it will cause confusion.
Another way, the legacy functions is a mistake indeed.

Reply:
Thanks for the quick response @beliefer - two questions:

1. I can't find any function descriptions that tell users that arrays are 0 based; could you link some? For example, here is the live documentation for [elementAt](https://spark.apache.org/docs/3.1.3/api/python/reference/api/pyspark.sql.functions.element_at.html), which uses 1 based indexing.
2. The SQL99 standard specifies that SQL Arrays use 1 based indexing. For example, PostGres uses 1 based index Arrays also. Even if we ignored all other legacy spark functions, is there a reason you prefer 0 based indexing?

Reply:
@beliefer @zhengruifeng I've amended the code to be 0 indexed now, and hence, exactly reproduce what snowflake are doing. Thank you for your prompt feedback.

Reply:
@Daniel-Davies I'm sorry. I said wrong. It seems the array functions supported by Spark starts at 1. So there are two ways:
1. All the new array functions still starts at 1. 
2. Another way, the legacy functions starts at 1 is a mistake indeed. For follow snowflake, we have to break change.

Reply:
No worries- I can revert to the last commit if we want to go with (1). My understanding is that standard SQL99 specifies that arrays are 1 indexed, which is why Spark arrays are 1 indexed.

Personally, (1) sounds like a good option to me. However I appreciate that this is out of my knowledge level, so I'm happy to pick whichever one of my commits that works for you.

Reply:
Please hold on. We are not get the final conclusion. cc @cloud-fan @gengliangwang @zhengruifeng Do you have an idea or preference ?

Reply:
thanks for the update.

> All array functions of Spark follow a 1-base index to my understanding

the old index were problematic, and we are switching to 0-indexed, see 
https://github.com/apache/spark/blob/4321604fe0b8f54b7cc5d1372f9b89a91fbad8b2/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala#L2172-L2195

Reply:
Thanks @zhengruifeng! In which case, this PR should be ready for review. Thank you for the update.

Reply:
We are not switching to 0-based index. The reason for adding `Get` is ANSI fallback: https://github.com/apache/spark/commit/b6cf3e4598fb6ae9f9ed28c7d5a0d4152453a669

Reply:
@cloud-fan thank you for the clarification

@Daniel-Davies  sorry for ask you to switch to 0-based index, my bad.